rfc/rfc-46/rfc-46.md

```diff
@@ -0,0 +1,159 @@
+   2. Plug-in point for user-defined combination semantics
+3. Gradually deprecate, phase-out and eventually remove `HoodieRecordPayload` abstraction
+
+Phasing out usage of `HoodieRecordPayload` will also bring the benefit of avoiding to use Java reflection in the hot-path, which
+is known to have poor performance (compared to non-reflection based instantiation).
+
+#### Combine API Engine
+
+Stateless component interface providing for API Combining Records will look like following:
+
+```java
+interface HoodieRecordCombiningEngine {
+  
+  default HoodieRecord precombine(HoodieRecord older, HoodieRecord newer) {
+    if (spark) {
+      precombineSpark((SparkHoodieRecord) older, (SparkHoodieRecord) newer);
+    } else if (flink) {
+      // precombine for Flink
+    }
+  }
+
+   /**
+    * Spark-specific implementation 
+    */
+  SparkHoodieRecord precombineSpark(SparkHoodieRecord older, SparkHoodieRecord newer);
+  
+  // ...
+}
+```
+Where user can provide their own subclass implementing such interface for the engines of interest.
+
+#### Migration from `HoodieRecordPayload` to `HoodieRecordCombiningEngine`
+
+To warrant backward-compatibility (BWC) on the code-level with already created subclasses of `HoodieRecordPayload` currently
+already used in production by Hudi users, we will provide a BWC-bridge in the form of instance of `HoodieRecordCombiningEngine`, that will 
+be using user-defined subclass of `HoodieRecordPayload` to combine the records.
+
+Leveraging such bridge will make provide for seamless BWC migration to the 0.11 release, however will be removing the performance 
+benefit of this refactoring, since it would unavoidably have to perform conversion to intermediate representation (Avro). To realize
+full-suite of benefits of this refactoring, users will have to migrate their merging logic out of `HoodieRecordPayload` subclass and into
+new `HoodieRecordCombiningEngine` implementation.
+
+### Refactoring Flows Directly Interacting w/ Records:
+
+As was called out prior to achieve the goal of being able to sustain engine-internal representations being held by `HoodieRecord` 
+class w/o compromising major components' neutrality (ie being engine-agnostic), such components directly interacting w/
+records' payloads today will have to be refactored to instead interact w/ standardized `HoodieRecord`s API.
+
+Following major components will be refactored:
+
+1. `WriteHandle`s will be  
+   1. Accepting `HoodieRecord` instead of raw Avro payload (avoiding Avro conversion)
+   2. Using Combining API engine to merge records (when necessary) 
+   3. Passes `HoodieRecord` as is to `FileWriter`
+2. `FileWriter`s will be 
+   1. Accepting `HoodieRecord`
+   2. Will be engine-specific (so that they're able to handle internal record representation)
+3. `RecordReader`s 
+   1. API will be returning opaque `HoodieRecord` instead of raw Avro payload
```

Comment:
was trying to look for the exact classes: are they `HoodieWriteHandle`, `HoodieFileWriter` and `HoodieFileReader` ?

Reply:
Correct

Reply:
ok would you fix the names so it reflects the exact class names?